lnwire/signature_test.go

```diff
@@ -24,67 +25,80 @@ func TestSignatureSerializeDeserialize(t *testing.T) {
-				": %s, %s", e.S, e2.S)
+		if !e.IsEqual(e2) {
+			return fmt.Errorf("pre/post-serialize sigs don't "+
+				"match: %x, %x", e.Serialize(), e2.Serialize())
 		}
 		return nil
 	}
 
-	sig := btcec.Signature{}
-
 	// Check R = N-1, S = 128.
-	sig.R = big.NewInt(1) // Allocate a big.Int before we call .Sub.
-	sig.R.Sub(btcec.S256().N, sig.R)
-	sig.S = big.NewInt(128)
+	r := big.NewInt(1) // Allocate a big.Int before we call .Sub.
+	r.Sub(btcec.S256().N, r)
+	rScalar := new(btcec.ModNScalar)
+	rScalar.SetByteSlice(r.Bytes())
+
+	sig := ecdsa.NewSignature(rScalar, new(btcec.ModNScalar).SetInt(128))
 	err := signatureSerializeDeserialize(sig)
 	if err != nil {
 		t.Fatalf("R = N-1, S = 128: %s", err.Error())
 	}
 
 	// Check R = N-1, S = 127.
-	sig.S = big.NewInt(127)
+	sig = ecdsa.NewSignature(rScalar, new(btcec.ModNScalar).SetInt(127))
 	err = signatureSerializeDeserialize(sig)
 	if err != nil {
 		t.Fatalf("R = N-1, S = 127: %s", err.Error())
 	}
 
 	// Check R = N-1, S = N>>1.
-	sig.S.Set(btcec.S256().N)
-	sig.S.Rsh(sig.S, 1)
+	s := new(big.Int).Set(btcec.S256().N)
+	s.Rsh(s, 1)
+	sScalar := new(btcec.ModNScalar)
+	sScalar.SetByteSlice(s.Bytes())
+	sig = ecdsa.NewSignature(rScalar, sScalar)
 	err = signatureSerializeDeserialize(sig)
 	if err != nil {
 		t.Fatalf("R = N-1, S = N>>1: %s", err.Error())
 	}
 
 	// Check R = N-1, S = N.
-	sig.S.Set(btcec.S256().N)
+	s = new(big.Int).Set(btcec.S256().N)
+	overflow := sScalar.SetByteSlice(s.Bytes())
+	if !overflow {
+		t.Fatalf("Expect ModNScalar to overflow when setting N but " +
+			"didn't")
+	}
+	sig = ecdsa.NewSignature(rScalar, sScalar)
 	err = signatureSerializeDeserialize(sig)
-	if err.Error() != "signature S isn't 1 or more" {
+	if err.Error() != "invalid signature: S is 0" {
 		t.Fatalf("R = N-1, S = N should become R = N-1, S = 0: %s",
 			err.Error())
 	}
 
 	// Check R = N-1, S = N-1.
-	sig.S.Sub(sig.S, big.NewInt(1))
-	err = signatureSerializeDeserialize(sig)
-	if err.Error() != "Pre/post-serialize Ss don't match: 115792089237316"+
-		"195423570985008687907852837564279074904382605163141518161494"+
-		"336, 1" {
-		t.Fatalf("R = N-1, S = N-1 should become R = N-1, S = 1: %s",
-			err.Error())
-	}
+	//s = new(big.Int).Set(btcec.S256().N)
+	//s.Sub(s, big.NewInt(1))
+	//sScalar.SetByteSlice(s.Bytes())
+	//sig = ecdsa.NewSignature(rScalar, sScalar)
+	//err = signatureSerializeDeserialize(sig)
+	//if err.Error() != "Pre/post-serialize Ss don't match: 115792089237316"+
+	//	"195423570985008687907852837564279074904382605163141518161494"+
+	//	"336, 1" {
+	//	t.Fatalf("R = N-1, S = N-1 should become R = N-1, S = 1: %s",
+	//		err.Error())
+	//}
 
 	// Check R = 2N, S = 128
-	sig.R.Mul(btcec.S256().N, big.NewInt(2))
-	sig.S.Set(big.NewInt(127))
-	err = signatureSerializeDeserialize(sig)
-	if err.Error() != "R is over 32 bytes long without padding" {
-		t.Fatalf("R = 2N, S = 128, R should be over 32 bytes: %s",
-			err.Error())
-	}
+	//r = new(big.Int).Set(btcec.S256().N)
```

Comment:
We can't overflow anymore because we're using ModNScalar. Should we keep this test case and try to replicate it in a different way? Or just add a test case for _de-serializing_ an invalid value?

Reply:
If you pass something that's too large, then `SetByteSlice` will return an error, so that can be checked.

Reply:
You mean the `overflow` boolean? Yeah, I added that as a check indeed.